File: c++/longest_substring_with_at_most_k_distinct_characters.cpp
/*
Example
For example, Given s = "eceba", k = 3,
T is "eceb" which its length is 4.

Challenge
O(n), n is the size of the string s.
*/

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    /**
     * Maintain a sliding window starting from the first
     * character. Use a hashmap to record every character's
     * frequency in the window. Each time, increment the window's
     * right index by one and insert the character just coming into
     * the window into the hashmap. If the hashmap's size is greater than
     * k, it means there are more than k distince characters, and
     * we need to increment window's left index till the map's size
     * is equal to k.
     */
    int longestSubstring(const string& str, int k) {
        if (k < 1) return 0;
        unordered_map<char, int> charCountMap;
        int left = 0;
        int right = left;
        int maxLen = 1;
        while (right < str.size()) {
            auto it = charCountMap.find(str[right]);
            if (it != charCountMap.end()) {
                ++it->second;
            } else {
                charCountMap[str[right]] = 1;
            }

            while (charCountMap.size() > k) {
                it = charCountMap.find(str[left++]);
                if (it->second == 1) charCountMap.erase(it);
                else it->second--;
            }

            maxLen = max(maxLen, right - left + 1);
            ++right;
        }
        return maxLen;
    }
};

int main()
{
    string str = "ecebaaaaaaaccccccdbbbbb";
    int k = 2;
    Solution sol;
    int length = sol.longestSubstring(str, k);
    cout << "Length = " << length << endl;
    return 0;
}

File: java/LongestSubstringWithAtMostKDistinctCharacters.java
/**
 * Created on 5/25/16.
 * Given a string, find the length of the longest substring T that contains at most k distinct characters.
 *
 * For example, Given s = “eceba” and k = 2,
 *
 * T is "ece" which its length is 3.
 */

public class LongestSubstringWithAtMostKDistinctCharacters {

    public static void main(String[] args) {
        String s = "ecebaaaaaaaccccccdbbbbb";
        int k = 2;
        LongestSubstringWithAtMostKDistinctCharacters app = new LongestSubstringWithAtMostKDistinctCharacters();
        int maxLength = app.lengthOfLongestSubstringKDistinct(s, k);
        System.out.println("The max length is " + maxLength + " for string "
                + s + " with " + k + " distinct characters");
    }

    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        int[] charSet = new int[26];
        for (int i = 0; i < 26; ++i) charSet[i] = 0;
        int left = 0, right = 0;
        int count = 0; // count of distinct characters in the current window [left, right]
        int maxLength = 0;
        while (right < s.length()) {
            int index = s.charAt(right) - 'a';
            if (charSet[index] > 0) {
                ++charSet[index];
            } else { // one more distinct character entering the window
                charSet[index] = 1;
                ++count;
            }

            while (count > k) {
                int i = s.charAt(left) - 'a';
                if (charSet[i] == 1) {
                    --charSet[i];
                    --count;
                } else {
                    --charSet[i];
                }
                ++left;
            }

            maxLength = Math.max(right - left + 1, maxLength);
            ++right;
        }

        return maxLength;
    }
}
